spec.md

```diff
@@ -244,6 +244,58 @@ A common use of an externally defined module is to represent a Verilog module
 that will be written separately and provided together with FIRRTL-generated
 Verilog to downstream tools.
 
+## Functions
+
+A function has a name, list of arguments, and a list of statements representing
+a data-flow graph of stateless circuit components.  Each argument is specified 
+by its direction, which may be input or output, a name, and the data type of the 
+argument.
+
+Functions participate in width-inference (see [@sec:width-inference]) at 
+invocation sites in the same way modules to at instantiation sites.
+
+Functions cannot contain memories, registers, or instances.  This makes them
+"stateless"
+
+Functions can optional specify that they are "pure" functions.  These do not
+have side-effecting operations (such as printf).
+
+
+``` firrtl
+function MyPureFunction pure :
+  input foo : UInt
+  output bar : UInt
+  bar <= foo
+
+function MyFunction :
+  input clk: Clock
+  input halt: UInt<1>
+  stop(clk, halt, 42) : optional_name
+```
+
+## Intrinsics
+
+An intrinsic is a function. Intrinsics have all the components of a function
+except  a list of statements.  An intrinsic is a firrtl implementation provided 
+function.  The name of intrinsics are implementation defined.  A firrtl 
+implementation shall reject IR with unknown intrinsics.  Being functions,
+intrinsics are stateless.
+
+The widths of all intrinsic ports must be specified.  Width inference, described 
+in [@sec:width-inference], is not supported.
+
+Intrinsics can be "pure".  These do not have side-effects.  Whether a particular 
+intrinsic is pure or not is implementation defined and implementations shall 
+check that intrinsics declarations have the expected arguments, types, and 
+purity.
+
+``` firrtl
+intrinsic synopsys.4mux.u8 pure :
```

Comment:
does a .fir file need to include this `intrinsic` in it if it plans to use it?

Reply:
I think that's the idea. This way the FIR file declares what it expects the intrinsic to be in terms of type signature and name, so the FIR file remains self-contained. And the compiler would just verify that the expected type signature matches what the actual internal intrinsic wants.

Reply:
in that sense is an `intrinsic` the same as an `external function`? Should we use the word `extfunction` to parallel the term `extmodule`?

Reply:
I'd expect there to be a difference. The `intrinsic` is something the compiler understands and will process/expand during compilation, so it doesn't make its way to the output. An `extfunction` I'd expect to be transported all the way through and to be emitted for example as a DPI call in the SV output.

Reply:
why do we need a special word for that, and only for functions? Maybe we want the same thing for `extmodule`s as well.

I guess I'm generally confused why we are treating "function" which is basically "module with no sequential logic in it" so differently/non-parallel.

Reply:
I vote that these are different.  External functions might be useful, `extfunction` makes sense to me for that, but I think should be separate from intrinsics (IMHO).  Functions are written by users, intrinsics are by compiler.

Are functions kinda like macros now? And/or do they lower to something suitable in verilog?

Presently we're able to compile a FIRRTL file w/`extmodule`'s to verilog in a reasonable way, not sure what we'd do for `extfunction`s (but maybe this is straightforward?).   Bringing it up so we're don't accidentally cross over the line into "you have to link this before you can produce verilog".

Reply:
extfunction is user-supplied, intrinsics are implementation provided and known.  They are just extfunctions, but they are ones the compiler directly knows.

Reply:
this was answered in this comment: https://github.com/chipsalliance/firrtl-spec/pull/43#discussion_r975619940